File: crates/yokan/demo/shooter.py
# /// script
# requires-python = ">=3.14"
# ///
"""Pyxel Shooter, ported to Yokan.

The original is `09_shooter.py` from Pyxel's examples (Takashi Kitao,
MIT, https://github.com/kitao/pyxel), and the sprite sheet in
`assets/shooter.png` is that example's own two 8x8 sprites, written out
with Pyxel's palette. The port follows it line by line: `pyxel.rect`
becomes `rect`, `pyxel.btn` becomes `keys.down`, `pyxel.blt` becomes
`sprite`, and the color numbers are the same numbers, because inside a
canvas a color is an index into the palette this file declares.

What is different, and why. Speeds that were fractional (1.5 px a
frame) are carried in tenths of a pixel and drawn whole, because a
pixel grid has no half pixels. The music and the sound effects are
gone: there is no audio here yet. So are the gamepad and `pyxel.quit`.
Everything else — three scenes, a hundred parallax stars, the enemy
that sways as it falls, rectangle collisions, expanding blasts — is
the game.

Arrow keys move, space fires, enter starts and restarts.
"""
import os
import random
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from yokan import (  # noqa: E402
    canvas,
    circle,
    circle_outline,
    column,
    every,
    keys,
    pixel,
    pixel_text,
    rect,
    run,
    sprite,
    store,
    value,
)

WIDTH = 120
HEIGHT = 160

SCENE_TITLE = 0
SCENE_PLAY = 1
SCENE_GAMEOVER = 2

NUM_STARS = 100
STAR_COLOR_HIGH = 12
STAR_COLOR_LOW = 5

PLAYER_WIDTH = 8
PLAYER_HEIGHT = 8
PLAYER_SPEED = 2

BULLET_WIDTH = 2
BULLET_HEIGHT = 8
BULLET_COLOR = 11
BULLET_SPEED = 4

ENEMY_WIDTH = 8
ENEMY_HEIGHT = 8
# Pyxel's 1.5 px a frame, in tenths.
ENEMY_SPEED = 15

BLAST_START_RADIUS = 1
BLAST_END_RADIUS = 8
BLAST_COLOR_IN = 7
BLAST_COLOR_OUT = 10

SHEET = "demo/assets/shooter.png"


@value
class Star:
    x: int
    # `y` is what the canvas draws; `y10` is where the star really is.
    y: int
    y10: int
    speed10: int
    col: int


@value
class Bullet:
    x: int
    y: int


@value
class Enemy:
    x: int
    y: int
    x10: int
    y10: int
    flip: bool
    offset: int


@value
class Blast:
    x: int
    y: int
    radius: int


@store
class Game:
    scene: int = 0
    score: int = 0
    frame: int = 0
    title_col: int = 0
    px: int = 56
    py: int = 140
    stars: list[Star] = []
    bullets: list[Bullet] = []
    enemies: list[Enemy] = []
    blasts: list[Blast] = []
    # Pyxel's own sixteen colors, which is what makes the numbers in
    # this file mean what they mean in the original.
    palette: list[str] = [
        "#000000", "#2b335f", "#7e2072", "#19959c",
        "#8b4852", "#395c98", "#a9c1ff", "#eeeeee",
        "#d4186c", "#d38441", "#e9c35b", "#70c6a9",
        "#7696de", "#a3a3a3", "#ff9798", "#edc7b0",
    ]

    def boot(self) -> None:
        random.seed(7)
        out: list[Star] = []
        for i in range(NUM_STARS):
            x = random.randint(0, WIDTH - 1)
            y = random.randint(0, HEIGHT - 1)
            speed10 = random.randint(10, 25)
            col = STAR_COLOR_LOW
            if speed10 > 18:
                col = STAR_COLOR_HIGH
            out = out + [Star(x, y, y * 10, speed10, col)]
        self.stars = out

    def tick(self) -> None:
        self.frame = self.frame + 1
        self.title_col = self.frame % 16
        Game.move_stars()
        if self.scene == SCENE_TITLE:
            if keys.pressed("enter"):
                self.scene = SCENE_PLAY
        elif self.scene == SCENE_PLAY:
            Game.play()
        else:
            Game.over()

    def move_stars(self) -> None:
        out: list[Star] = []
        for s in self.stars:
            y10 = s.y10 + s.speed10
            if y10 >= HEIGHT * 10:
                y10 = y10 - HEIGHT * 10
            out = out + [Star(s.x, y10 // 10, y10, s.speed10, s.col)]
        self.stars = out

    def play(self) -> None:
        if self.frame % 6 == 0:
            x = random.randint(0, WIDTH - ENEMY_WIDTH)
            self.enemies = self.enemies + [
                Enemy(x, 0, x * 10, 0, False, random.randint(0, 59))
            ]
        Game.collide()
        Game.move_player()
        Game.move_bullets()
        Game.move_enemies()
        Game.move_blasts()

    def over(self) -> None:
        Game.move_bullets()
        Game.move_enemies()
        Game.move_blasts()
        if keys.pressed("enter"):
            self.scene = SCENE_PLAY
            self.px = 56
            self.py = 140
            self.score = 0
            self.enemies = []
            self.bullets = []
            self.blasts = []

    def move_player(self) -> None:
        x = self.px
        y = self.py
        if keys.down("left"):
            x = x - PLAYER_SPEED
        if keys.down("right"):
            x = x + PLAYER_SPEED
        if keys.down("up"):
            y = y - PLAYER_SPEED
        if keys.down("down"):
            y = y + PLAYER_SPEED
        self.px = min(max(x, 0), WIDTH - PLAYER_WIDTH)
        self.py = min(max(y, 0), HEIGHT - PLAYER_HEIGHT)
        if keys.pressed("space"):
            self.bullets = self.bullets + [
                Bullet(self.px + 3, self.py - 4)
            ]

    def move_bullets(self) -> None:
        out: list[Bullet] = []
        for b in self.bullets:
            y = b.y - BULLET_SPEED
            if y + BULLET_HEIGHT - 1 >= 0:
                out = out + [Bullet(b.x, y)]
        self.bullets = out

    def move_enemies(self) -> None:
        out: list[Enemy] = []
        for e in self.enemies:
            x10 = e.x10
            flip = True
            if (self.frame + e.offset) % 60 < 30:
                x10 = x10 + ENEMY_SPEED
                flip = False
            else:
                x10 = x10 - ENEMY_SPEED
            y10 = e.y10 + ENEMY_SPEED
            if y10 <= (HEIGHT - 1) * 10:
                out = out + [
                    Enemy(x10 // 10, y10 // 10, x10, y10, flip, e.offset)
                ]
        self.enemies = out

    def move_blasts(self) -> None:
        out: list[Blast] = []
        for b in self.blasts:
            r = b.radius + 1
            if r <= BLAST_END_RADIUS:
                out = out + [Blast(b.x, b.y, r)]
        self.blasts = out

    def collide(self) -> None:
        """The two rectangle tests, resolved into new lists.

        A value is not edited in place, so where the original sets
        `is_alive = False` and filters afterwards, this keeps the ones
        that live."""
        live_enemies: list[Enemy] = []
        hit: list[int] = []
        blasts_ = self.blasts
        score_ = self.score
        struck_player = False
        for e in self.enemies:
            struck = False
            for i in range(len(self.bullets)):
                b = self.bullets[i]
                if (
                    e.x + ENEMY_WIDTH > b.x
                    and b.x + BULLET_WIDTH > e.x
                    and e.y + ENEMY_HEIGHT > b.y
                    and b.y + BULLET_HEIGHT > e.y
                ):
                    struck = True
                    hit = hit + [i]
            if struck:
                blasts_ = blasts_ + [
                    Blast(e.x + 4, e.y + 4, BLAST_START_RADIUS)
                ]
                score_ = score_ + 10
            elif (
                self.px + PLAYER_WIDTH > e.x
                and e.x + ENEMY_WIDTH > self.px
                and self.py + PLAYER_HEIGHT > e.y
                and e.y + ENEMY_HEIGHT > self.py
            ):
                blasts_ = blasts_ + [
                    Blast(self.px + 4, self.py + 4, BLAST_START_RADIUS)
                ]
                struck_player = True
            else:
                live_enemies = live_enemies + [e]
        live_bullets: list[Bullet] = []
        for i in range(len(self.bullets)):
            if i in hit:
                continue
            live_bullets = live_bullets + [self.bullets[i]]
        self.enemies = live_enemies
        self.bullets = live_bullets
        self.blasts = blasts_
        self.score = score_
        if struck_player:
            self.scene = SCENE_GAMEOVER


every(0.033, Game.tick)


def view():
    with column(spacing=0, padding=0):
        with canvas(WIDTH, HEIGHT, scale=4, background=0, palette=Game.palette):
            for s in Game.stars:
                pixel(s.x, s.y, s.col)
            if Game.scene == SCENE_TITLE:
                pixel_text(35, 66, "Pyxel Shooter", Game.title_col)
                pixel_text(31, 126, "- PRESS ENTER -", 13)
            elif Game.scene == SCENE_PLAY:
                sprite(Game.px, Game.py, SHEET, 0, 0, PLAYER_WIDTH, PLAYER_HEIGHT, colkey=0)
            else:
                pixel_text(43, 66, "GAME OVER", 8)
                pixel_text(31, 126, "- PRESS ENTER -", 13)
            for b in Game.bullets:
                rect(b.x, b.y, BULLET_WIDTH, BULLET_HEIGHT, BULLET_COLOR)
            for e in Game.enemies:
                sprite(e.x, e.y, SHEET, 8, 0, ENEMY_WIDTH, ENEMY_HEIGHT, colkey=0, flip_x=e.flip)
            for bl in Game.blasts:
                circle(bl.x, bl.y, bl.radius, BLAST_COLOR_IN)
                circle_outline(bl.x, bl.y, bl.radius, BLAST_COLOR_OUT)
            pixel_text(39, 4, f"SCORE {Game.score:5}", 7)


if __name__ == "__main__":
    run(view, title="Pyxel Shooter", width=480.0, height=640.0, on_start=Game.boot)

File: gomamochi/demo/shooter.go
// Pyxel Shooter, ported to Gomamochi.
//
// The original is `shooter.py` from Pyxel's examples (Takashi Kitao,
// MIT, https://github.com/kitao/pyxel), and `assets/shooter.png` is that
// example's own image bank written out with Pyxel's palette.
//
// What is different, and why. The effects are WAV files rather than
// the original's chiptune, since the engine plays files; a run under a
// script is silent, so the gate still compares two silent runs. And the
// numbers come from a generator written here rather than from a random
// source, so that a reader can follow every enemy to its place.
//
// Arrows move, space fires, enter starts and restarts, q closes.
package main

import (
	"fmt"

	. "github.com/i2y/yokan/gomamochi"
)

const (
	width  = 120
	height = 160

	sndShoot = "demo/assets/sound/shoot.wav"
	sndBlast = "demo/assets/sound/blast.wav"
	sndOver  = "demo/assets/sound/over.wav"

	sceneTitle    = 0
	scenePlay     = 1
	sceneGameOver = 2

	numStars      = 100
	starColorHigh = 12
	starColorLow  = 5

	playerWidth  = 8
	playerHeight = 8
	playerSpeed  = 2

	bulletWidth  = 2
	bulletHeight = 8
	bulletColor  = 11
	bulletSpeed  = 4

	enemyWidth  = 8
	enemyHeight = 8
	// Pyxel's 1.5 px a frame, in tenths.
	enemySpeed = 15

	blastStartRadius = 1
	blastEndRadius   = 8
	blastColorIn     = 7
	blastColorOut    = 10

	sheet = "demo/assets/shooter.png"
)

// Pyxel's own sixteen colors, which is what makes the numbers in this
// file mean what they mean in the original.
var palette = []string{
	"#000000", "#2b335f", "#7e2072", "#19959c",
	"#8b4852", "#395c98", "#a9c1ff", "#eeeeee",
	"#d4186c", "#d38441", "#e9c35b", "#70c6a9",
	"#7696de", "#a3a3a3", "#ff9798", "#edc7b0",
}

// Whole numbers from arithmetic alone, so both runs play one game.
type Roll struct{ s int }

func (r *Roll) next(lo, hi int) int {
	r.s = (r.s*1103515245 + 12345) % 2147483648
	return lo + ((r.s >> 8) % (hi - lo + 1))
}

// Star: `y` is what the canvas draws; `y10` is where the star really is.
type Star struct{ x, y, y10, speed10, col int }

type Bullet struct{ x, y int }

type Enemy struct {
	x, y, x10, y10 int
	flip           bool
	offset         int
}

type Blast struct{ x, y, radius int }

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// The original's `//` rounds toward minus infinity, and an enemy
// sweeping off the left edge has a negative tenth.
func div10(v int) int {
	if v < 0 && v%10 != 0 {
		return v/10 - 1
	}
	return v / 10
}

type Game struct {
	scene    int
	score    int
	frame    int
	titleCol int
	px, py   int
	stars    []Star
	bullets  []Bullet
	enemies  []Enemy
	blasts   []Blast
	roll     *Roll
}

func newGame() *Game {
	g := &Game{px: 56, py: 140, roll: &Roll{s: 7}}
	for i := 0; i < numStars; i++ {
		g.bootStar()
	}
	return g
}

func (g *Game) bootStar() {
	x := g.roll.next(0, width-1)
	y := g.roll.next(0, height-1)
	speed10 := g.roll.next(10, 25)
	col := starColorLow
	if speed10 > 18 {
		col = starColorHigh
	}
	g.stars = append(g.stars, Star{x, y, y * 10, speed10, col})
}

func (g *Game) tick() {
	if KeyPressed("q") {
		Quit()
	}
	g.frame += 1
	g.titleCol = g.frame % 16
	g.moveStars()
	switch g.scene {
	case sceneTitle:
		if KeyPressed("enter") {
			g.scene = scenePlay
		}
	case scenePlay:
		g.play()
	default:
		g.over()
	}
}

func (g *Game) moveStars() {
	out := make([]Star, 0, len(g.stars))
	for _, s := range g.stars {
		y10 := s.y10 + s.speed10
		if y10 >= height*10 {
			y10 -= height * 10
		}
		out = append(out, Star{s.x, y10 / 10, y10, s.speed10, s.col})
	}
	g.stars = out
}

func (g *Game) play() {
	if g.frame%6 == 0 {
		x := g.roll.next(0, width-enemyWidth)
		g.enemies = append(g.enemies, Enemy{x, 0, x * 10, 0, false, g.roll.next(0, 59)})
	}
	g.collide()
	g.movePlayer()
	g.moveBullets()
	g.moveEnemies()
	g.moveBlasts()
}

func (g *Game) over() {
	g.moveBullets()
	g.moveEnemies()
	g.moveBlasts()
	if !KeyPressed("enter") {
		return
	}
	g.scene = scenePlay
	g.px = 56
	g.py = 140
	g.score = 0
	g.enemies = nil
	g.bullets = nil
	g.blasts = nil
}

func (g *Game) movePlayer() {
	x, y := g.px, g.py
	if KeyDown("left") {
		x -= playerSpeed
	}
	if KeyDown("right") {
		x += playerSpeed
	}
	if KeyDown("up") {
		y -= playerSpeed
	}
	if KeyDown("down") {
		y += playerSpeed
	}
	g.px = minInt(maxInt(x, 0), width-playerWidth)
	g.py = minInt(maxInt(y, 0), height-playerHeight)
	if !KeyPressed("space") {
		return
	}
	g.bullets = append(g.bullets, Bullet{g.px + 3, g.py - 4})
	AudioPlay(sndShoot, 0.35)
}

func (g *Game) moveBullets() {
	out := make([]Bullet, 0, len(g.bullets))
	for _, b := range g.bullets {
		y := b.y - bulletSpeed
		if y+bulletHeight-1 >= 0 {
			out = append(out, Bullet{b.x, y})
		}
	}
	g.bullets = out
}

func (g *Game) moveEnemies() {
	out := make([]Enemy, 0, len(g.enemies))
	for _, e := range g.enemies {
		x10 := e.x10
		flip := true
		if (g.frame+e.offset)%60 < 30 {
			x10 += enemySpeed
			flip = false
		} else {
			x10 -= enemySpeed
		}
		y10 := e.y10 + enemySpeed
		if y10 <= (height-1)*10 {
			out = append(out, Enemy{div10(x10), y10 / 10, x10, y10, flip, e.offset})
		}
	}
	g.enemies = out
}

func (g *Game) moveBlasts() {
	out := make([]Blast, 0, len(g.blasts))
	for _, b := range g.blasts {
		r := b.radius + 1
		if r <= blastEndRadius {
			out = append(out, Blast{b.x, b.y, r})
		}
	}
	g.blasts = out
}

// The two rectangle tests, resolved into new lists. Where the original
// sets `is_alive = False` and filters afterwards, this keeps the ones
// that live.
func (g *Game) collide() {
	live := make([]Enemy, 0, len(g.enemies))
	hit := make([]bool, len(g.bullets))
	struck := false
	for _, e := range g.enemies {
		enemyStruck := false
		for i, b := range g.bullets {
			if shot(e, b) {
				enemyStruck = true
				hit[i] = true
			}
		}
		if enemyStruck {
			g.blasts = append(g.blasts, Blast{e.x + 4, e.y + 4, blastStartRadius})
			g.score += 10
			AudioPlay(sndBlast, 0.5)
		} else if g.rammed(e) {
			g.blasts = append(g.blasts, Blast{g.px + 4, g.py + 4, blastStartRadius})
			struck = true
			AudioPlay(sndOver, 0.6)
		} else {
			live = append(live, e)
		}
	}
	kept := make([]Bullet, 0, len(g.bullets))
	for i, b := range g.bullets {
		if !hit[i] {
			kept = append(kept, b)
		}
	}
	g.enemies = live
	g.bullets = kept
	if struck {
		g.scene = sceneGameOver
	}
}

func shot(e Enemy, b Bullet) bool {
	return e.x+enemyWidth > b.x && b.x+bulletWidth > e.x &&
		e.y+enemyHeight > b.y && b.y+bulletHeight > e.y
}

func (g *Game) rammed(e Enemy) bool {
	return g.px+playerWidth > e.x && e.x+enemyWidth > g.px &&
		g.py+playerHeight > e.y && e.y+enemyHeight > g.py
}

func (g *Game) View() Element {
	return Column(
		Canvas(width, height).Scale(4).Background(0).Palette(palette...).Paint(func(p *Painter) {
			for _, s := range g.stars {
				p.Pixel(s.x, s.y, s.col)
			}
			switch g.scene {
			case sceneTitle:
				p.PixelText(35, 66, "Pyxel Shooter", g.titleCol)
				p.PixelText(31, 126, "- PRESS ENTER -", 13)
			case scenePlay:
				p.Sprite(g.px, g.py, sheet, 0, 0, playerWidth, playerHeight, 0, false, false)
			default:
				p.PixelText(43, 66, "GAME OVER", 8)
				p.PixelText(31, 126, "- PRESS ENTER -", 13)
			}
			for _, b := range g.bullets {
				p.Rect(b.x, b.y, bulletWidth, bulletHeight, bulletColor)
			}
			for _, e := range g.enemies {
				p.Sprite(e.x, e.y, sheet, 8, 0, enemyWidth, enemyHeight, 0, e.flip, false)
			}
			for _, b := range g.blasts {
				p.Circle(b.x, b.y, b.radius, blastColorIn)
				p.CircleOutline(b.x, b.y, b.radius, blastColorOut)
			}
			p.PixelText(39, 4, fmt.Sprintf("SCORE %5d", g.score), 7)
		}),
	).Spacing(0).Padding(0)
}

func main() {
	app := newGame()
	Every(0.033, func() { app.tick() })
	// Padding 0: the canvas IS the app, so it paints to the window's
	// edge rather than sitting inside the engine's ring.
	Run(app, Title("Pyxel Shooter"), Size(480, 640), Padding(0))
}
